spectrocloud/cluster_common_virtual_machine.go

```diff
@@ -404,36 +398,90 @@ func toSpecCreateRequest(d *schema.ResourceData) *models.V1ClusterVirtualMachine
+	}
 	return vmSpec
 }
 
-func toVirtualMachineUpdateRequest(d *schema.ResourceData, vm *models.V1ClusterVirtualMachine) (bool, *models.V1ClusterVirtualMachine, error) {
+func toVirtualMachineUpdateRequest(d *schema.ResourceData, vm *models.V1ClusterVirtualMachine) (bool, bool, *models.V1ClusterVirtualMachine, error) {
 	requireUpdate := false
-	if d.HasChange("name") {
-		vm.Metadata.Name = d.Get("name").(string)
-		requireUpdate = true
-	}
-	if d.HasChange("namespace") {
-		vm.Metadata.Namespace = d.Get("namespace").(string)
-		requireUpdate = true
-	}
+	needRestart := false
 	if d.HasChange("cpu_cores") {
 		vm.Spec.Template.Spec.Domain.CPU.Cores = int64(d.Get("cpu_cores").(int))
 		requireUpdate = true
+		needRestart = true
 	}
 	if d.HasChange("memory") {
 		vm.Spec.Template.Spec.Domain.Resources.Requests = map[string]interface{}{
 			"memory": d.Get("memory").(string),
 		}
 		requireUpdate = true
+		needRestart = true
 	}
 	if _, ok := d.GetOk("image_url"); ok && d.HasChange("image_url") {
 		vm.Metadata.Namespace = d.Get("namespace").(string)
 		requireUpdate = true
+		needRestart = true
+	}
+	if _, ok := d.GetOk("labels"); ok && d.HasChange("labels") {
+		vm.Metadata.Labels = toVMLabels(d)
+		requireUpdate = true
+	}
+	if _, ok := d.GetOk("annotations"); ok && d.HasChange("annotations") {
+		vm.Metadata.Annotations = toVMUpdateAnnotations(vm.Metadata.Annotations, d)
+		requireUpdate = true
+	}
+	if _, ok := d.GetOk("volume_spec"); ok && d.HasChange("volume_spec") {
+		vm.Spec.Template.Spec.Volumes = prepareVolumeSpec(d)
+		requireUpdate = true
+		needRestart = true
+	}
+	if _, ok := d.GetOk("network_spec"); ok && d.HasChange("network_spec") {
+		vm.Spec.Template.Spec.Networks = prepareNetworkSpec(d)
+		requireUpdate = true
+		needRestart = true
+	}
+	if _, ok := d.GetOk("devices"); ok && d.HasChange("devices") {
+		vm.Spec.Template.Spec.Domain.Devices.Disks, vm.Spec.Template.Spec.Domain.Devices.Interfaces = prepareDevices(d)
+		requireUpdate = true
+		needRestart = true
+	}
+	if run, ok := d.GetOk("run_on_launch"); ok && d.HasChange("run_on_launch") {
+		vm.Spec.Running = run.(bool)
+		if run.(bool) == true {
+			vm.Spec.RunStrategy = ""
+		} else {
+			vm.Spec.RunStrategy = "Manual"
+		}
+
 	}
 
 	// There is issue in Ally side, team asked as to explicitly make deletion-time to nil before put operation, after fix will remove.
 	vm.Spec.Template.Metadata.DeletionTimestamp = nil
 	vm.Metadata.DeletionTimestamp = nil
-	return requireUpdate, vm, nil
+	return requireUpdate, needRestart, vm, nil
+}
+
+func toVMAnnotations(d *schema.ResourceData) map[string]string {
+	annotation := make(map[string]string)
+	if _, ok := d.GetOk("annotations"); ok {
+		for k, a := range d.Get("annotations").(map[string]interface{}) {
+			annotation[k] = a.(string)
+		}
+		return annotation
+	} else {
+		return nil
+	}
+}
+
+func toVMUpdateAnnotations(existingAnnotation map[string]string, d *schema.ResourceData) map[string]string {
```

Comment:
can we reuse annotations above and from other resources too.

Reply:
In VM annotations are a little different, here during VM launch, we have a few default annotation sets.
during updation and addition, we need to handle this separately hence I have introduced a new function.